rmw_fastrtps_shared_cpp/src/custom_publisher_info.cpp

```diff
@@ -160,6 +197,40 @@ void PubListener::set_on_new_event_callback(
   }
 }
 
+void
+PubListener::on_publication_matched(
+  eprosima::fastdds::dds::DataWriter * /* writer */,
+  const eprosima::fastdds::dds::PublicationMatchedStatus & info)
+{
+  rmw_event_type_t event_type = RMW_EVENT_INVALID;
+  {
+    std::lock_guard<std::mutex> lock(discovery_m_);
+    if (info.current_count_change == 1) {
+      subscriptions_.insert(eprosima::fastrtps::rtps::iHandle2GUID(info.last_subscription_handle));
+      event_type = RMW_EVENT_PUBLICATION_MATCHED;
+    } else if (info.current_count_change == -1) {
+      subscriptions_.erase(eprosima::fastrtps::rtps::iHandle2GUID(info.last_subscription_handle));
+      event_type = RMW_EVENT_PUBLICATION_UNMATCHED;
+    }
+  }
+
+  if (event_type != RMW_EVENT_INVALID) {
+    {
+      std::lock_guard<std::mutex> lock_mutex(on_new_event_m_);
+
+      matched_status_.total_count_change += info.total_count_change;
+      matched_status_.current_count = info.current_count;
+      matched_status_.current_count_change += info.current_count_change;
+
+      unmatched_status_.total_count_change += info.total_count_change;
+      unmatched_status_.current_count = info.current_count;
+      unmatched_status_.current_count_change += info.current_count_change;
```

Comment:
these are stored only if event_type is equal to RMW_EVENT_PUBLICATION_UNMATCHED?